Firestore/core/src/firebase/firestore/model/delete_mutation.cc

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/model/delete_mutation.h"
+
+#include <cstdlib>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/model/document.h"
+#include "Firestore/core/src/firebase/firestore/model/field_path.h"
+#include "Firestore/core/src/firebase/firestore/model/field_value.h"
+#include "Firestore/core/src/firebase/firestore/model/no_document.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+static_assert(
+    sizeof(Mutation) == sizeof(DeleteMutation),
+    "DeleteMutation may not have additional members (everything goes in Rep)");
+
+DeleteMutation::DeleteMutation(DocumentKey key, Precondition precondition)
+    : Mutation(std::make_shared<Rep>(std::move(key), std::move(precondition))) {
+}
+
+DeleteMutation::DeleteMutation(const Mutation& mutation) : Mutation(mutation) {
+  HARD_ASSERT(type() == Type::Delete);
```

Comment:
FWIW- It occurs to me that this means casting an invalid instance will result in a nullptr dereference.  Is that OK?

Reply:
If we put the check in `rep()` and make sure `type()` uses the `rep()` getter, then that would work.